web/server/go.md

```diff
@@ -0,0 +1,498 @@
+allowing mutability of underlying infrastructure
+and enforcing code reuse.
+
+Let's say our Owner can leave the application,
+and in this case the owner's dogs should also be marked as inactive.
+We want our Dog model code to maintain this constraint.
+
+Here's a method to demonstrate that relationship:
+
+```go
+package models
+
+type Dog struct {
+    // ...
+    Status Status
+}
+
+type Owner struct {
+    Dogs []Dog
+    Status Status
+}
+
+func (o *Owner) Deactivate() {
+    o.Status = Inactive
+    dogs := o.Dogs[:0]
+    for _, dog := range o.Dogs {
+        dog.Status = Inactive
+        dogs = append(dogs, dog)
+    }
+    o.Dogs = dogs
+}
+```
+
+The purpose here is to set up the public model API
+to only allow valid modifications throughout the codebase.
+We've explicitly linked the dog and owner
+to enforce any data constraints.
+Similar methodology may follow for reactivating the owner,
+changing a dog's owner, etc.
+Note that in many cases when writing Go code,
+we'd set parameters like these as unexported
+to avoid invalid state changes,
+however this isn't possible due to required exposure
+by DB/JSON packages that will use the model's public fields.
+
+We've continued to separate any DB concerns
+to outside the model.
+For example,
+it's likely we'd want to wrap these changes in a transaction:
+
+```go
+package services
+
+func DeactiveOwner(owner Owner) error {
+    transaction := db.Transaction.Begin()
+    owner.Deactivate()
+    err := transaction.Save(owner)
+    if err !=nil {
+        transaction.Rollback()
+        return err
+    }
+    err = transaction.Save(owner.Dogs)
+    if err !=nil {
+        transaction.Rollback()
+        return err
+    }
+    err := transaction.Save(owner)
+    transaction.Commit()
+    return nil
+}
+```
+
+Separating the database concerns from the model
+allows us to interact with them separate of the data structure concerns.
+For example,
+if we wanted to use a locking feature specific to Postgres
+or our database connector library,
+we can make those changes independent of the model,
+allowing the model to be reused elsewhere
+without pulling in possibly irrelevant packages or patterns.
+It also allows for isolating experiments and changes
+to a small portion of the codebase,
+rather than propagated them throughout all model usages.
+
+#### Where/When to Write Query Code
+
+The above examples make use of an abstracted dependency for a DB connector.
+In reality,
+any ORM or query interface is likely to be much more complex than `Fetch`/`Save`.
+As data requirements become more complex,
+it becomes much harder to maintain models as purely structural.
+It's hard to make an exact recommendation for Go code structure
+as data requirements may vary considerably.
+Backends could vary from REST APIs,
+highly normalized SQL tables,
+serialized document stores,
+sparsely column oriented data,
+and so on.
+
+Some general considerations when organizing query code are:
```

Comment:
Would be cool to build out some links to real world examples here in the future.